### PR TITLE
#2518. support metadata transmission in http.

### PR DESCRIPTION
**Pull request by cvictory**
# Description

In some situation, the metadata cannot be transported to sdk in http api.
For example the response of output binding have metadata, but it don't transport to app.

I think there are two ways to support the feature:
1. add the metadata into httpHeader.
2.marshal the return body that both contain metadata and body.

Maybe the 2nd way is the better one, but it will cause the sdk failure right now.
So I define some function in the http/response.go to support this feature that transfer metadata by HttpHeader.

## Issue reference

https://github.com/dapr/dapr/issues/2518

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
Metadata should be returned in the header, as you've done in this PR.

**Comment by yaron2** [comment]
@mukundansundar please review this.

**Comment by mukundansundar** [comment]
@halspang can you check if the e2e test for middleware is flaky? https://github.com/dapr/dapr/pull/2519/checks?check_run_id=1532370345 it failed this check alone and the other 3 e2e tests passed in KinD

File: pkg/http/responses.go [hunk 1]
```diff
@@ -41,13 +47,33 @@ func respond(ctx *fasthttp.RequestCtx, code int, obj []byte) {
 	}
 }
 
+// respond sets a default application/json content type if content type is not present
```
Comment (mukundansundar) [inline pkg/http/responses.go:50]: change comment.

File: pkg/http/responses.go [hunk 2]
```diff
@@ -41,13 +47,33 @@ func respond(ctx *fasthttp.RequestCtx, code int, obj []byte) {
...
+}
+
 // respondWithETaggedJSON overrides the content-type with application/json and etag header
 func respondWithETaggedJSON(ctx *fasthttp.RequestCtx, code int, obj []byte, etag string) {
 	respond(ctx, code, obj)
 	ctx.Response.Header.SetContentType(jsonContentTypeHeader)
 	ctx.Response.Header.Set(etagHeader, etag)
 }
 
+// respondWithETaggedJSON overrides the content-type with application/json and etag header
```
Comment (mukundansundar) [inline pkg/http/responses.go:70]: change comment.

**Comment by cvictory** [comment]
@mukundansundar  Hello. In the issue(https://github.com/dapr/dapr/issues/2531) , I suggest to introduce gRpc server Streaming or Bidirectional Streaming to implements pub/sub or some action like watch state.
And If in this way, we will have some difficult at http1.  

I want to know the which situation we should use http1?  what is the problem If we don't maintain the http1?

**Comment by artursouza** [comment]
> @mukundansundar Hello. In the issue(#2531) , I suggest to introduce gRpc server Streaming or Bidirectional Streaming to implements pub/sub or some action like watch state.
> And If in this way, we will have some difficult at http1.
> 
> I want to know the which situation we should use http1? what is the problem If we don't maintain the http1?


We want to keep http 1 support for now. On the other hand, this should be a discussion outside this PR.

Regarding this PR, we are OK to merge this after addressing the comments. I am closing this due to inactivity, please, reopen or cut a new PR if these changes are intended to be merged.

**Pull request closed without merging**